### ISSUE TITLE
longtail training

### ISSUE THREAD

**Issue by ZehuaLiuY** [ZehuaLiuY/Species-Classification#10]
在你的代码中，你可以通过实现以下方法来结合 **Logit Adjustment (LA)**、**Class-Balanced Focal Loss (CB)** 和 **Weight Balancing (WB)**，以便在微调模型中使用这三种长尾识别（LTR）技术。

---

### 1. **Logit Adjustment (LA)**

Logit Adjustment 通过调整模型输出的 logits 来平衡类别分布。你需要修改交叉熵损失函数以引入类别频率的影响。

#### 实现方法：

```python
import torch.nn.functional as F

class LogitAdjustedCrossEntropyLoss(torch.nn.Module):
    def __init__(self, class_freq, tau=1.0):
        super().__init__()
        self.class_freq = class_freq
        self.tau = tau  # Temperature scaling factor

    def forward(self, logits, labels):
        logit_adjustment = torch.log(self.class_freq + 1e-6) * self.tau
        adjusted_logits = logits - logit_adjustment.to(logits.device)
        return F.cross_entropy(adjusted_logits, labels)
```

#### 在代码中使用：
你需要计算每个类别的频率：

```python
# 计算类别频率
class_counts = np.bincount(train_dataset.indices)
class_freq = class_counts / sum(class_counts)
class_freq_tensor = torch.tensor(class_freq, dtype=torch.float32)

# 使用 LogitAdjustedCrossEntropyLoss
criterion = LogitAdjustedCrossEntropyLoss(class_freq=class_freq_tensor, tau=1.0)
```

---

### 2. **Class-Balanced Focal Loss (CB)**

CB 是一种结合类别平衡和 Focal Loss 的方法，适合处理类别不平衡问题。

#### 实现方法：

```python
class ClassBalancedFocalLoss(torch.nn.Module):
    def __init__(self, class_freq, gamma=2.0, beta=0.99):
        super().__init__()
        effective_num = 1.0 - np.power(beta, class_freq)
        weights = (1.0 - beta) / np.array(effective_num)
        weights = weights / np.sum(weights)
        self.class_weights = torch.tensor(weights, dtype=torch.float32)
        self.gamma = gamma

    def forward(self, logits, labels):
        alpha = self.class_weights.to(logits.device)[labels]
        probs = torch.softmax(logits, dim=-1)
        target_probs = probs[range(len(labels)), labels]
        focal_weight = torch.pow(1.0 - target_probs, self.gamma)
        loss = -alpha * focal_weight * torch.log(target_probs + 1e-8)
        return loss.mean()
```

#### 在代码中使用：
与 Logit Adjustment 类似，计算 `class_freq` 后使用：

```python
criterion = ClassBalancedFocalLoss(class_freq=class_freq_tensor.numpy(), gamma=2.0, beta=0.99)
```

---

### 3. **Weight Balancing (WB)**

Weight Balancing 通过调整最后一层分类器的权重大小来平衡类别。

#### 实现方法：

在模型定义中，对最后一层权重添加约束：

```python
def apply_weight_balancing(model, max_norm=1.0):
    for name, param in model.named_parameters():
        if 'classifier.weight' in name:  # 假设分类器的权重在 classifier.weight
            param.data = torch.renorm(param.data, p=2, dim=0, maxnorm=max_norm)
```

#### 在代码中使用：

在每次优化步骤之后，手动调用：

```python
optimizer.step()
apply_weight_balancing(model, max_norm=1.0)
```

---

### 整合到训练代码

将上述方法嵌入到训练流程中：

```python
# Initialize criterion with the desired LTR method
if args.loss == "logit_adjustment":
    criterion = LogitAdjustedCrossEntropyLoss(class_freq=class_freq_tensor, tau=1.0)
elif args.loss == "class_balanced_focal":
    criterion = ClassBalancedFocalLoss(class_freq=class_freq_tensor.numpy(), gamma=2.0, beta=0.99)
else:
    criterion = torch.nn.CrossEntropyLoss()

# Training loop
for epoch in range(1, args.epochs + 1):
    train_epoch_loss, train_epoch_acc, global_step = train_one_epoch(
        model, train_loader, optimizer, criterion, device, epoch, global_step_start=global_step
    )

    if args.loss == "weight_balancing":
        apply_weight_balancing(model, max_norm=1.0)
```

---

### 参数选择与实验对比

- 在运行脚本时通过命令行参数选择不同的 LTR 方法：
```bash
python train.py --loss logit_adjustment
python train.py --loss class_balanced_focal
python train.py --loss weight_balancing
```

- 通过日志记录和验证集评估模型性能，选择最佳方法。

---

通过以上方法，你可以方便地集成 Logit Adjustment、Class-Balanced Focal Loss 和 Weight Balancing，提升模型对长尾类别的识别能力。
